File: utils/src/test/java/org/robolectric/util/inject/InjectorTest.java
package org.robolectric.util.inject;

import static com.google.common.truth.Truth.assertThat;
import static org.junit.Assert.fail;

import com.google.auto.service.AutoService;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;
import javax.annotation.Priority;
import javax.inject.Inject;
import javax.inject.Named;
import org.junit.Before;
import org.junit.Ignore;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

@RunWith(JUnit4.class)
public class InjectorTest {

  private Injector.Builder builder;
  private Injector injector;
  private final List<Class<?>> pluginClasses = new ArrayList<>();

  @Before
  public void setUp() throws Exception {
    builder = new Injector.Builder();
    injector = builder.build();
  }

  @Test
  public void whenImplSpecified_shouldProvideInstance() throws Exception {
    injector = builder.bind(Thing.class, MyThing.class).build();

    assertThat(injector.getInstance(Thing.class))
        .isInstanceOf(MyThing.class);
  }

  @Test
  public void whenImplSpecified_shouldUseSameInstance() throws Exception {
    injector = builder.bind(Thing.class, MyThing.class).build();

    Thing thing = injector.getInstance(Thing.class);
    assertThat(injector.getInstance(Thing.class)).isSameInstanceAs(thing);
  }

  @Test
  public void whenServiceSpecified_shouldProvideInstance() throws Exception {
    assertThat(injector.getInstance(Thing.class))
        .isInstanceOf(ThingFromServiceConfig.class);
  }

  @Test
  public void whenServiceSpecified_shouldUseSameInstance() throws Exception {
    Thing thing = injector.getInstance(Thing.class);
    assertThat(injector.getInstance(Thing.class)).isSameInstanceAs(thing);
  }

  @Test
  public void whenConcreteClassRequested_shouldProvideInstance() throws Exception {
    assertThat(injector.getInstance(MyUmm.class))
        .isInstanceOf(MyUmm.class);
  }

  @Test
  public void whenDefaultSpecified_shouldProvideInstance() throws Exception {
    injector = builder.bindDefault(Umm.class, MyUmm.class).build();

    assertThat(injector.getInstance(Umm.class))
        .isInstanceOf(MyUmm.class);
  }

  @Test
  public void whenDefaultSpecified_shouldUseSameInstance() throws Exception {
    Thing thing = injector.getInstance(Thing.class);
    assertThat(injector.getInstance(Thing.class)).isSameInstanceAs(thing);
  }

  @Test
  public void whenNoImplOrServiceOrDefaultSpecified_shouldThrow() throws Exception {
    try {
      injector.getInstance(Umm.class);
      fail();
    } catch (InjectionException e) {
      // ok
    }
  }

  @Test
  public void registerDefaultService_providesFallbackImplOnlyIfNoServiceSpecified()
      throws Exception {
    builder.bindDefault(Thing.class, MyThing.class);

    assertThat(injector.getInstance(Thing.class))
        .isInstanceOf(ThingFromServiceConfig.class);

    builder.bindDefault(Umm.class, MyUmm.class);
    assertThat(injector.getInstance(Thing.class))
        .isInstanceOf(ThingFromServiceConfig.class);
  }

  @Test
  public void shouldPreferSingularPublicConstructorAnnotatedInject() throws Exception {
    injector = builder
        .bind(Thing.class, MyThing.class)
        .bind(Umm.class, MyUmm.class)
        .build();

    Umm umm = injector.getInstance(Umm.class);
    assertThat(umm).isNotNull();
    assertThat(umm).isInstanceOf(MyUmm.class);

    MyUmm myUmm = (MyUmm) umm;
    assertThat(myUmm.thing).isNotNull();
    assertThat(myUmm.thing).isInstanceOf(MyThing.class);

    assertThat(myUmm.thing).isSameInstanceAs(injector.getInstance(Thing.class));
  }

  @Test
  public void shouldAcceptSingularPublicConstructorWithoutInjectAnnotation() throws Exception {
    injector = builder
        .bind(Thing.class, MyThing.class)
        .bind(Umm.class, MyUmmNoInject.class)
        .build();

    Umm umm = injector.getInstance(Umm.class);
    assertThat(umm).isNotNull();
    assertThat(umm).isInstanceOf(MyUmmNoInject.class);

    MyUmmNoInject myUmm = (MyUmmNoInject) umm;
    assertThat(myUmm.thing).isNotNull();
    assertThat(myUmm.thing).isInstanceOf(MyThing.class);

    assertThat(myUmm.thing).isSameInstanceAs(injector.getInstance(Thing.class));
  }

  @Test
  public void whenArrayRequested_mayReturnMultiplePlugins() throws Exception {
    MultiThing[] multiThings = injector.getInstance(MultiThing[].class);

    // X comes first because it has a higher priority
    assertThat(classesOf(multiThings))
        .containsExactly(MultiThingX.class, MultiThingA.class).inOrder();
  }

  @Test
  public void whenCollectionRequested_mayReturnMultiplePlugins() throws Exception {
    ThingRequiringMultiThings it = injector.getInstance(ThingRequiringMultiThings.class);

    // X comes first because it has a higher priority
    assertThat(classesOf(it.multiThings))
        .containsExactly(MultiThingX.class, MultiThingA.class).inOrder();
  }

  @Test
  public void whenListRequested_itIsUnmodifiable() throws Exception {
    ThingRequiringMultiThings it = injector.getInstance(ThingRequiringMultiThings.class);

    try {
      it.multiThings.clear();
      fail();
    } catch (Exception e) {
      assertThat(e).isInstanceOf(UnsupportedOperationException.class);
    }
  }

  @Test public void autoFactory_factoryMethodsCreateNewInstances() throws Exception {
    injector = builder.bind(Umm.class, MyUmm.class).build();
    FooFactory factory = injector.getInstance(FooFactory.class);
    Foo chauncey = factory.create("Chauncey");
    assertThat(chauncey.name).isEqualTo("Chauncey");

    Foo anotherChauncey = factory.create("Chauncey");
    assertThat(anotherChauncey).isNotSameInstanceAs(chauncey);
  }

  @Test public void autoFactory_injectedValuesComeFromSuperInjector() throws Exception {
    injector = builder.bind(Umm.class, MyUmm.class).build();
    FooFactory factory = injector.getInstance(FooFactory.class);
    Foo chauncey = factory.create("Chauncey");
    assertThat(chauncey.thing).isSameInstanceAs(injector.getInstance(Thing.class));
  }

  @Test public void whenFactoryRequested_createsInjectedFactory() throws Exception {
    injector = builder.bind(Umm.class, MyUmm.class).build();
    FooFactory factory = injector.getInstance(FooFactory.class);
    Foo chauncey = factory.create("Chauncey");
    assertThat(chauncey.name).isEqualTo("Chauncey");

    Foo anotherChauncey = factory.create("Chauncey");
    assertThat(anotherChauncey).isNotSameInstanceAs(chauncey);

    assertThat(chauncey.thing).isSameInstanceAs(injector.getInstance(Thing.class));
  }

  @Test public void scopedInjector_shouldCheckParentBeforeProvidingDefault() throws Exception {
    injector = builder.build();
    Injector subInjector = new Injector.Builder(injector).build();

    MyUmm subUmm = subInjector.getInstance(MyUmm.class);
    assertThat(injector.getInstance(MyUmm.class)).isSameInstanceAs(subUmm);
  }

  @Test public void shouldInjectByNamedKeys() throws Exception {
    injector = builder
        .bind(new Injector.Key<>(String.class, "namedThing"), "named value")
        .bind(String.class, "unnamed value")
        .build();
    NamedParams namedParams = injector.getInstance(NamedParams.class);
    assertThat(namedParams.withName).isEqualTo("named value");
    assertThat(namedParams.withoutName).isEqualTo("unnamed value");
  }

  @Test public void shouldPreferPluginsOverConcreteClass() throws Exception {
    PluginFinder pluginFinder = new PluginFinder(new MyServiceFinderAdapter(pluginClasses));
    Injector injector = new Injector.Builder(null, pluginFinder).build();
    pluginClasses.add(SubclassOfConcreteThing.class);
    ConcreteThing instance = injector.getInstance(ConcreteThing.class);
    assertThat(instance.getClass()).isEqualTo(SubclassOfConcreteThing.class);
  }

  @Test
  public void subInjectorIsUsedForResolvingTransitiveDependencies() throws Exception {
    FakeSandboxManager sandboxManager = injector.getInstance(FakeSandboxManager.class);
    FakeSdk runtimeSdk = new FakeSdk("runtime");
    FakeSdk compileSdk = new FakeSdk("compile");
    FakeSandbox sandbox = sandboxManager.getSandbox(runtimeSdk, compileSdk);
    assertThat(sandbox.runtimeSdk).isSameInstanceAs(runtimeSdk);
    assertThat(sandbox.compileSdk).isSameInstanceAs(compileSdk);
  }

  @Test @Ignore("todo")
  public void objectsCreatedByFactoryShareTransitiveDependencies() throws Exception {
    FakeSandboxManager sandboxManager = injector.getInstance(FakeSandboxManager.class);
    FakeSdk runtimeSdk = new FakeSdk("runtime");
    FakeSdk compileASdk = new FakeSdk("compileA");
    FakeSdk compileBSdk = new FakeSdk("compileB");
    FakeSandbox sandboxA = sandboxManager.getSandbox(runtimeSdk, compileASdk);
    FakeSandbox sandboxB = sandboxManager.getSandbox(runtimeSdk, compileBSdk);
    assertThat(sandboxA.sandboxClassLoader).isSameInstanceAs(sandboxB.sandboxClassLoader);
  }

  @Test
  public void shouldProvideDecentErrorMessages() throws Exception {
    FakeSandboxManager sandboxManager = injector.getInstance(FakeSandboxManager.class);
    Exception actualException = null;
    try {
      sandboxManager.brokenGetSandbox();
      fail();
    } catch (Exception e) {
      actualException = e;
    }
    assertThat(actualException.getMessage())
        .contains("Failed to resolve dependency: FakeSandbox/FakeSdk/String");
  }

  @Test @Ignore("todo")
  public void shouldOnlyAttemptToResolveTypesKnownToClassLoader() throws Exception {
  }

  /////////////////////////////

  private List<? extends Class<?>> classesOf(Object[] items) {
    return classesOf(Arrays.asList(items));
  }

  private List<? extends Class<?>> classesOf(List<?> items) {
    return items.stream().map(Object::getClass).collect(Collectors.toList());
  }

  /** A thing. */
  public interface Thing {
  }

  public static class MyThing implements Thing {
  }

  public static class ConcreteThing {
  }

  public static class SubclassOfConcreteThing extends ConcreteThing {
  }

  /** Class for test. */
  @AutoService(Thing.class)
  public static class ThingFromServiceConfig implements Thing {
  }

  private interface Umm {

  }

  public static class MyUmm implements Umm {

    private final Thing thing;

    @Inject
    public MyUmm(Thing thing) {
      this.thing = thing;
    }

    @SuppressWarnings("unused")
    public MyUmm(String thingz) {
      this.thing = null;
    }
  }

  /** Class for test. */
  public static class MyUmmNoInject implements Umm {

    private final Thing thing;

    public MyUmmNoInject(Thing thing) {
      this.thing = thing;
    }
  }

  private interface MultiThing {

  }

  /** Class for test. */
  @Priority(-5)
  @AutoService(MultiThing.class)
  public static class MultiThingA implements MultiThing {
  }

  /** Class for test. */
  @AutoService(MultiThing.class)
  public static class MultiThingX implements MultiThing {
  }

  /** Class for test. */
  public static class ThingRequiringMultiThings {

    private List<MultiThing> multiThings;

    public ThingRequiringMultiThings(List<MultiThing> multiThings) {
      this.multiThings = multiThings;
    }
  }

  static class Foo {

    private final Thing thing;
    private final Umm umm;
    private final String name;

    public Foo(Thing thing, Umm umm, String name) {
      this.thing = thing;
      this.umm = umm;
      this.name = name;
    }
  }

  @AutoFactory
  interface FooFactory {
    Foo create(String name);
  }

  static class NamedParams {

    private final Thing thing;
    private final String withName;
    private final String withoutName;

    public NamedParams(Thing thing, @Named("namedThing") String withName, String withoutName) {
      this.thing = thing;
      this.withName = withName;
      this.withoutName = withoutName;
    }
  }

  static class FakeSdk {
    private final String name;

    public FakeSdk(String name) {
      this.name = name;
    }
  }

  static class FakeSandbox {

    private final FakeSdk runtimeSdk;
    private final FakeSdk compileSdk;
    private final FakeSandboxClassLoader sandboxClassLoader;

    public FakeSandbox(
        @Named("runtimeSdk") FakeSdk runtimeSdk,
        @Named("compileSdk") FakeSdk compileSdk,
        FakeSandboxClassLoader sandboxClassLoader) {
      this.runtimeSdk = runtimeSdk;
      this.compileSdk = compileSdk;
      this.sandboxClassLoader = sandboxClassLoader;
    }
  }

  static class FakeSandboxClassLoader {
    private final FakeSdk runtimeSdk;

    public FakeSandboxClassLoader(@Named("runtimeSdk") FakeSdk runtimeSdk) {
      this.runtimeSdk = runtimeSdk;
    }
  }

  static class FakeSandboxManager {

    private final FakeSandboxFactory sandboxFactory;

    public FakeSandboxManager(FakeSandboxFactory sandboxFactory) {
      this.sandboxFactory = sandboxFactory;
    }

    public FakeSandbox getSandbox(FakeSdk runtimeSdk, FakeSdk compileSdk) {
      return sandboxFactory.createSandbox(runtimeSdk, compileSdk);
    }

    public FakeSandbox brokenGetSandbox() {
      return sandboxFactory.createSandbox();
    }
  }

  @AutoFactory
  private interface FakeSandboxFactory {
    FakeSandbox createSandbox(@Named("runtimeSdk") FakeSdk runtimeSdk,
        @Named("compileSdk") FakeSdk compileSdk);
    FakeSandbox createSandbox();
  }
}

File: utils/src/test/java/org/robolectric/util/inject/InjectorTest.kt
package org.robolectric.util.inject

import com.google.auto.service.AutoService
import com.google.common.truth.Truth.assertThat
import java.util.*
import java.util.stream.Collectors
import javax.inject.Inject
import javax.inject.Named
import org.junit.Assert
import org.junit.Before
import org.junit.Ignore
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import org.robolectric.util.inject.*
import javax.annotation.Priority

@RunWith(JUnit4::class)
class InjectorTest {

  private lateinit var builder: Injector.Builder
  private lateinit var injector: Injector
  private val pluginClasses: MutableList<Class<*>> = ArrayList()

  @Before
  @Throws(Exception::class)
  fun setUp() {
    builder = Injector.Builder()
    injector = builder.build()
  }

  @Test
  @Throws(Exception::class)
  fun whenImplSpecified_shouldProvideInstance() {
    injector = builder.bind(Thing::class.java, MyThing::class.java).build()
    assertThat(injector.getInstance(Thing::class.java)).isInstanceOf(MyThing::class.java)
  }

  @Test
  @Throws(Exception::class)
  fun whenImplSpecified_shouldUseSameInstance() {
    injector = builder.bind(Thing::class.java, MyThing::class.java).build()
    val thing = injector.getInstance(Thing::class.java)
    assertThat(injector.getInstance(Thing::class.java)).isSameInstanceAs(thing)
  }

  @Test
  @Throws(Exception::class)
  fun whenServiceSpecified_shouldProvideInstance() {
    assertThat(injector.getInstance(Thing::class.java))
        .isInstanceOf(ThingFromServiceConfig::class.java)
  }

  @Test
  @Throws(Exception::class)
  fun whenServiceSpecified_shouldUseSameInstance() {
    val thing = injector.getInstance(Thing::class.java)
    assertThat(injector.getInstance(Thing::class.java)).isSameInstanceAs(thing)
  }

  @Test
  @Throws(Exception::class)
  fun whenConcreteClassRequested_shouldProvideInstance() {
    assertThat(injector.getInstance(MyUmm::class.java)).isInstanceOf(MyUmm::class.java)
  }

  @Test
  @Throws(Exception::class)
  fun whenDefaultSpecified_shouldProvideInstance() {
    injector = builder.bindDefault(Umm::class.java, MyUmm::class.java).build()
    assertThat(injector.getInstance(Umm::class.java)).isInstanceOf(MyUmm::class.java)
  }

  @Test
  @Throws(Exception::class)
  fun whenDefaultSpecified_shouldUseSameInstance() {
    val thing = injector.getInstance(Thing::class.java)
    assertThat(injector.getInstance(Thing::class.java)).isSameInstanceAs(thing)
  }

  @Test
  @Throws(Exception::class)
  fun whenNoImplOrServiceOrDefaultSpecified_shouldThrow() {
    try {
      injector.getInstance(Umm::class.java)
      Assert.fail()
    } catch (e: InjectionException) {
      // ok
    }
  }

  @Test
  @Throws(Exception::class)
  fun registerDefaultService_providesFallbackImplOnlyIfNoServiceSpecified() {
    builder.bindDefault(Thing::class.java, MyThing::class.java)
    assertThat(injector.getInstance(Thing::class.java))
        .isInstanceOf(ThingFromServiceConfig::class.java)
    builder.bindDefault(Umm::class.java, MyUmm::class.java)
    assertThat(injector.getInstance(Thing::class.java))
        .isInstanceOf(ThingFromServiceConfig::class.java)
  }

  @Test
  @Throws(Exception::class)
  fun shouldPreferSingularPublicConstructorAnnotatedInject() {
    injector =
        builder
            .bind(Thing::class.java, MyThing::class.java)
            .bind(Umm::class.java, MyUmm::class.java)
            .build()
    val umm = injector.getInstance(Umm::class.java)
    assertThat(umm).isNotNull()
    assertThat(umm).isInstanceOf(MyUmm::class.java)
    val myUmm = umm as MyUmm
    assertThat(myUmm.thing).isNotNull()
    assertThat(myUmm.thing).isInstanceOf(MyThing::class.java)
    assertThat(myUmm.thing).isSameInstanceAs(injector.getInstance(Thing::class.java))
  }

  @Test
  @Throws(Exception::class)
  fun shouldAcceptSingularPublicConstructorWithoutInjectAnnotation() {
    injector =
        builder
            .bind(Thing::class.java, MyThing::class.java)
            .bind(Umm::class.java, MyUmmNoInject::class.java)
            .build()
    val umm = injector.getInstance(Umm::class.java)
    assertThat(umm).isNotNull()
    assertThat(umm).isInstanceOf(MyUmmNoInject::class.java)
    val myUmm = umm as MyUmmNoInject
    assertThat(myUmm.thing).isNotNull()
    assertThat(myUmm.thing).isInstanceOf(MyThing::class.java)
    assertThat(myUmm.thing).isSameInstanceAs(injector.getInstance(Thing::class.java))
  }

  @Test
  @Throws(Exception::class)
  fun whenArrayRequested_mayReturnMultiplePlugins() {
    val multiThings: Array<MultiThing> = injector.getInstance(Array<MultiThing>::class.java)

    // X comes first because it has a higher priority
    assertThat(classesOf(multiThings))
        .containsExactly(MultiThingX::class.java, MultiThingA::class.java)
        .inOrder()
  }

  @Test
  @Throws(Exception::class)
  fun whenCollectionRequested_mayReturnMultiplePlugins() {
    val it = injector.getInstance(ThingRequiringMultiThings::class.java)

    // X comes first because it has a higher priority
    assertThat(classesOf(it.multiThings))
        .containsExactly(MultiThingX::class.java, MultiThingA::class.java)
        .inOrder()
  }

  @Test
  @Throws(Exception::class)
  fun whenListRequested_itIsUnmodifiable() {
    val it = injector.getInstance(ThingRequiringMultiThings::class.java)
    try {
      it.multiThings.clear()
      Assert.fail()
    } catch (e: Exception) {
      assertThat(e).isInstanceOf(UnsupportedOperationException::class.java)
    }
  }

  @Test
  @Throws(Exception::class)
  fun autoFactory_factoryMethodsCreateNewInstances() {
    injector = builder.bind(Umm::class.java, MyUmm::class.java).build()
    val factory = injector.getInstance(FooFactory::class.java)
    val chauncey = factory.create("Chauncey")
    assertThat(chauncey.name).isEqualTo("Chauncey")
    val anotherChauncey = factory.create("Chauncey")
    assertThat(anotherChauncey).isNotSameInstanceAs(chauncey)
  }

  @Test
  @Throws(Exception::class)
  fun autoFactory_injectedValuesComeFromSuperInjector() {
    injector = builder.bind(Umm::class.java, MyUmm::class.java).build()
    val factory = injector.getInstance(FooFactory::class.java)
    val chauncey = factory.create("Chauncey")
    assertThat(chauncey.thing).isSameInstanceAs(injector.getInstance(Thing::class.java))
  }

  @Test
  @Throws(Exception::class)
  fun whenFactoryRequested_createsInjectedFactory() {
    injector = builder.bind(Umm::class.java, MyUmm::class.java).build()
    val factory = injector.getInstance(FooFactory::class.java)
    val chauncey = factory.create("Chauncey")
    assertThat(chauncey.name).isEqualTo("Chauncey")
    val anotherChauncey = factory.create("Chauncey")
    assertThat(anotherChauncey).isNotSameInstanceAs(chauncey)
    assertThat(chauncey.thing).isSameInstanceAs(injector.getInstance(Thing::class.java))
  }

  @Test
  @Throws(Exception::class)
  fun scopedInjector_shouldCheckParentBeforeProvidingDefault() {
    injector = builder.build()
    val subInjector = Injector.Builder(injector).build()
    val subUmm = subInjector.getInstance(MyUmm::class.java)
    assertThat(injector.getInstance(MyUmm::class.java)).isSameInstanceAs(subUmm)
  }

  @Test
  @Throws(Exception::class)
  fun shouldInjectByNamedKeys() {
    injector =
        builder
            .bind(Injector.Key(String::class.java, "namedThing"), "named value")
            .bind(String::class.java, "unnamed value")
            .build()
    val namedParams = injector.getInstance(NamedParams::class.java)
    assertThat(namedParams.withName).isEqualTo("named value")
    assertThat(namedParams.withoutName).isEqualTo("unnamed value")
  }

  @Test
  @Throws(Exception::class)
  fun shouldPreferPluginsOverConcreteClass() {
    val pluginFinder = PluginFinder(MyServiceFinderAdapter(pluginClasses))
    val injector = Injector.Builder(null, pluginFinder).build()
    pluginClasses.add(SubclassOfConcreteThing::class.java)
    val instance = injector.getInstance(ConcreteThing::class.java)
    assertThat(instance.javaClass).isEqualTo(SubclassOfConcreteThing::class.java)
  }

  @Test
  @Throws(Exception::class)
  fun subInjectorIsUsedForResolvingTransitiveDependencies() {
    val sandboxManager = injector.getInstance(FakeSandboxManager::class.java)
    val runtimeSdk = FakeSdk("runtime")
    val compileSdk = FakeSdk("compile")
    val sandbox = sandboxManager.getSandbox(runtimeSdk, compileSdk)
    assertThat(sandbox.runtimeSdk).isSameInstanceAs(runtimeSdk)
    assertThat(sandbox.compileSdk).isSameInstanceAs(compileSdk)
  }

  @Test
  @Ignore("todo")
  @Throws(Exception::class)
  fun objectsCreatedByFactoryShareTransitiveDependencies() {
    val sandboxManager = injector.getInstance(FakeSandboxManager::class.java)
    val runtimeSdk = FakeSdk("runtime")
    val compileASdk = FakeSdk("compileA")
    val compileBSdk = FakeSdk("compileB")
    val sandboxA = sandboxManager.getSandbox(runtimeSdk, compileASdk)
    val sandboxB = sandboxManager.getSandbox(runtimeSdk, compileBSdk)
    assertThat(sandboxA.sandboxClassLoader).isSameInstanceAs(sandboxB.sandboxClassLoader)
  }

  @Test
  @Throws(Exception::class)
  fun shouldProvideDecentErrorMessages() {
    val sandboxManager = injector.getInstance(FakeSandboxManager::class.java)
    var actualException: Exception? = null
    try {
      sandboxManager.brokenGetSandbox()
      Assert.fail()
    } catch (e: Exception) {
      actualException = e
    }
    assertThat(actualException?.message)
        .contains("Failed to resolve dependency: FakeSandbox/FakeSdk/String")
  }

  @Test
  @Ignore("todo")
  @Throws(Exception::class)
  fun shouldOnlyAttemptToResolveTypesKnownToClassLoader() {}

  /////////////////////////////
  private fun classesOf(items: Array<*>): List<Class<*>?> {
    return classesOf(listOf(*items))
  }

  private fun classesOf(items: List<*>): List<Class<*>?> {
    return items
        .stream()
        .map { obj: Any? -> obj?.javaClass?.javaClass }
        .collect(Collectors.toList())
  }

  class MyThing : Thing

  open class ConcreteThing

  class SubclassOfConcreteThing : ConcreteThing()

  /** Class for test. */
  @AutoService(Thing::class) class ThingFromServiceConfig : Thing

  class MyUmm : Umm {
    internal val thing: Thing?

    @Inject
    constructor(thing: Thing?) {
      this.thing = thing
    }

    @SuppressWarnings("unused")
    constructor(thingz: String?) {
      thing = null
    }
  }

  /** Class for test. */
  class MyUmmNoInject(internal val thing: Thing) : Umm

  interface MultiThing

  /** Class for test. */
  @Priority(-5) @AutoService(MultiThing::class)
  class MultiThingA : MultiThing

  /** Class for test. */
  @AutoService(MultiThing::class)
  class MultiThingX : MultiThing

  /** Class for test. */
  class ThingRequiringMultiThings(var multiThings: MutableList<MultiThing>)

  internal class Foo
  internal constructor(internal val thing: Thing, private val umm: Umm, internal val name: String)

  @AutoFactory
  internal interface FooFactory {
    fun create(name: String?): Foo
  }

  internal class NamedParams(
      private val thing: String,
      @param:Named("namedThing") internal val withName: String,
      internal val withoutName: String
  )

  internal class FakeSdk(private val name: String)

  internal class FakeSandbox(
    @param:Named("runtimeSdk") internal val runtimeSdk: FakeSdk,
    @param:Named("compileSdk") internal val compileSdk: FakeSdk,
    internal val sandboxClassLoader: FakeSandboxClassLoader
  )

  internal class FakeSandboxClassLoader(@param:Named("runtimeSdk") private val runtimeSdk: FakeSdk)

  internal class FakeSandboxManager(private var sandboxFactory: FakeSandboxFactory) {

    fun getSandbox(runtimeSdk: FakeSdk?, compileSdk: FakeSdk?): FakeSandbox {
      return sandboxFactory.createSandbox(runtimeSdk, compileSdk)
    }

    fun brokenGetSandbox(): FakeSandbox {
      return sandboxFactory.createSandbox()
    }
  }

  @AutoFactory
  internal interface FakeSandboxFactory {
    fun createSandbox(
      @Named("runtimeSdk") runtimeSdk: FakeSdk?,
      @Named("compileSdk") compileSdk: FakeSdk?
    ): FakeSandbox

    fun createSandbox(): FakeSandbox
  }
}
